crypto/rsa/rsa_local.h

```diff
@@ -29,6 +29,8 @@ DECLARE_ASN1_ITEM(RSA_PRIME_INFO)
 DEFINE_STACK_OF(RSA_PRIME_INFO)
 
 struct rsa_st {
+    OPENSSL_CTX *libctx;
```

Comment:
Here be dragons. I'm really unsure about this, and would rather have libctx passed to the functions that need it...

Reply:
Not sure why. We've done the same thing in lots of other places, e.g. EVP_PKEY_CTX, BN_CTX, CRYPTO_EX_DATA, EC_GROUP (I stopped looking after that - the list goes on). I don't see why this one would be any different.

Reply:
Its also in the dh and dsa structs.. Dragons be everywhere. Its ok.

Reply:
@slontis, that is incorrect, the DH and DSA do *not* contain a `OPENSSL_CTX` pointer.

Reply:
We do have `OPENSSL_CTX` pointers in some context structures, and that's ok, those are fairly short-lived structures that are meant to use for a specific operation, and it's quite all right that they are tied to a specific library context.

When it comes to key structures (which I view as quite different from context structures), I'm not so sure.  I'll write a discussion point not tied to the diff.

Reply:
I was thinking of PROV_DSA_CTX & PROV_DH_CTX (in PR #10826)

Reply:
I hope it's clear by now that I have no problem whatsoever with an `OPENSSL_CTX` pointer in those context structure.

Reply:
It is not straight forward to "pass the libctx to functions that need it". For example `rsa_ossl_public_decrypt` requires the libctx, but is an implementation of `rsa_pub_dec` from RSA_METHOD. Changing the signature of these functions is a breaking change. There are possible workarounds involving adding new functions to RSA_METHOD - but that's really nasty and messy and we should avoid doing this if at all possible. Adding the libctx to the RSA structure is simple and straight forward and we've done exactly that in many of the structures that I've previously mentioned. I see no reason why we can't do the same here (in fact I can't think of any downside at all - so I'm not sure where your objection is coming from).

Reply:
"many other structures" mostly being context structures, which I view as a different kind of beast.
As for `rsa_ossl_public_decrypt`, it's perfectly possible to make a variant of it that takes an extra `OPENSSL_CTX` parameter, and have `rsa_ossl_public_decrypt` call that with NULL.  This is also something we do all over the place.

Reply:
Believe you me, I do understand the luring easier way, and there are many place where we do that, but I feel that for these structures, which are payload and long lived rather than context and fairly short lived, we're walking onto dangerous grounds.  Also, like I try to discuss separately from the diff, there are moments where we suddenly have two libctx to deal with, one being in the provider operation context (what we often call provctx) and one being in the key, and I dread the moment we lose control and they are suddenly different.

Reply:
I'm still sceptical re the necessity for this, but figured that it's not as bad as I first believed.  So ok, I guess we'll see down the line what this gives...